### DISCUSSION TITLE
Define/Unset Distinct Attribute At Search Time

### DISCUSSION THREAD

**Discussion by gmourier** [meilisearch/product#414] [Feedback & Feature Proposal]
Following: https://github.com/meilisearch/meilisearch/discussions/2221

[geoffatsource](https://github.com/geoffatsource) asked this question in [Q&A](https://github.com/meilisearch/meilisearch/discussions/categories/q-a)

I want to be able to set an index to have a distinct attribute, but be able to ignore it on certain types of queries (for instance geo or looking up all items in a distinct set)... is there a way to fashion a specific query to ignore the distinct setting?

---

Meilisearch could expose a `disableDistinctAttribute` boolean on `/search` APIs to solve it! 

Feel free to upvote this feature proposal, tell us more about your use case, or suggest another solution! ✨

**Comment by Kerollmops** [comment]
It is technically possible if I remember correctly.
There is nothing required for an attribute to be used as a distinct attribute besides the fact that it must be faceted.

> Meilisearch could expose a `disableDistinctAttribute` boolean on `/search` APIs to solve it!

Or we could have an `attributesForDistinct` in the settings and be able to define the one we want to use at search time (`/search`).

**Comment by gmourier** [comment]
Hey @geoffatsource 👋

I was thinking about this, and I have some questions about an idea.

1. The settings `distinctAttribute` could become `distinctAttributes` and host an array of attributes usable at search time.
2. At search time, it is possible to specify the distinct attribute to use among the defined `distinctAttributes` or no distinct attribute.

This would allow several things:

- Users would not have to change the settings multiple times when determining the `distinctAttribute` to use at search during their development phase. Thus speeding up the Meilisearch integration.
- Users will be able to use different `distinctAttribute` depending on the context of the search on the front end (They may want to have different implementations of the search around a web product). They are not tied to a single distinct attribute defined in the settings.
- They may find it interesting to propose a component allowing the user to choose among the list of `distinctAttributes` when interacting with the search components.

What do you think? Would you have an interest in seeing the capability implemented in this way?

**Reply by geoffatsource** [reply]
As long as it didn't harm performance significantly, I would totally use this :)

**Reply by gmourier** [reply]
Thanks, @geoffatsource!

**Reply by Rkallenkoot** [reply]
This sounds good! Also very clearly explained in the spec: https://specs.meilisearch.com/specifications/text/0032-distinct-attribute.html#_3-future-possibilities

**Comment by macraig** [comment]
Hey folks, we have some news to share! 

[v1.9](https://github.com/meilisearch/meilisearch/releases/tag/v1.9.0) has been released and it includes setting `distinct` attributes at search time.

Happy searching ✨
